sc-memory/sc-memory/sc_keynodes.hpp

```diff
@@ -6,153 +6,302 @@
+   */
+  template <typename T1, typename T2, typename T3>
+  _SC_EXTERN ScTemplateKeynode Triple(T1 const & param1, T2 const & param2, T3 const & param3) noexcept;
+
+  /*!
+   * @brief Appends provided quintuple to sc-template keynode.
+   * @note This method can be used in the same way as a method `Quintuple` of the class `ScTemplate`.
+   */
+  template <typename T1, typename T2, typename T3, typename T4, typename T5>
+  _SC_EXTERN ScTemplateKeynode
+  Quintuple(T1 const & param1, T2 const & param2, T3 const & param3, T4 const & param4, T5 const & param5) noexcept;
 
-  SC_PROPERTY(Keynode("binary_int64"), ForceCreate(ScType::NodeConstClass))
-  _SC_EXTERN static ScAddr kBinaryInt64;
+private:
+  std::list<std::function<void(ScTemplate &)>> m_constructionInitializers;
 
-  SC_PROPERTY(Keynode("binary_uint8"), ForceCreate(ScType::NodeConstClass))
-  _SC_EXTERN static ScAddr kBinaryUInt8;
+  template <typename T>
+  static auto HandleParam(T const & param) noexcept;
 
-  SC_PROPERTY(Keynode("binary_uint16"), ForceCreate(ScType::NodeConstClass))
-  _SC_EXTERN static ScAddr kBinaryUInt16;
+  /*!
+   * @brief Initializes the keynode with the given context and memory structure address.
+   * @param context A sc-memory context used for initialization.
+   */
+  void Initialize(ScMemoryContext * context) override;
+};
 
-  SC_PROPERTY(Keynode("binary_uint32"), ForceCreate(ScType::NodeConstClass))
-  _SC_EXTERN static ScAddr kBinaryUInt32;
+#include "sc_keynodes.tpp"
+
+/*!
+ * @class ScKeynodes
+ * @brief A base class for keynodes declaration. It's like a namespace. Use it as a base class for own keynodes.
+ * @note This class already contains frequently used keynodes. Use them in your code.
+ *
+ * @code
+ * // File my_keynodes.hpp:
+ * #pragma once
+ *
+ * #include <sc-memory/sc_keynodes.hpp>
+ *
+ * class MyKeynodes : public ScKeynodes
+ * {
+ * public:
+ *   static inline ScKeynode const my_keynode_class_a{"my_keynode_class_a", ScType::NodeConstClass};
+ *   static inline ScKeynode const my_keynode_relation_b{"my_keynode_relation_b", ScType::NodeConstNoRole};
+ * };
+ * @endcode
+ */
+class _SC_EXTERN ScKeynodes : public ScObject
+{
+  friend class ScMemory;
+  friend class ScModule;
+  friend class ScLink;
 
-  SC_PROPERTY(Keynode("binary_uint64"), ForceCreate(ScType::NodeConstClass))
-  _SC_EXTERN static ScAddr kBinaryUInt64;
+public:
+  _SC_EXTERN static inline ScKeynode const myself{"myself"};
+
+  _SC_EXTERN static inline ScKeynode const nrel_inclusion{"nrel_inclusion", ScType::NodeConstNoRole};
+  _SC_EXTERN static inline ScKeynode const rrel_1{"rrel_1", ScType::NodeConstRole};
+  _SC_EXTERN static inline ScKeynode const rrel_2{"rrel_2", ScType::NodeConstRole};
+  _SC_EXTERN static inline ScKeynode const rrel_3{"rrel_3", ScType::NodeConstRole};
+  _SC_EXTERN static inline ScKeynode const nrel_basic_sequence{"nrel_basic_sequence", ScType::NodeConstNoRole};
```

Comment:
not all keynodes from old coreKeynodes were moved here, for example rrel_4 is missing